README
***********
NSCA README
***********

General
-------
This version of NSCA has been forked and modified for usage with Icinga.
License and Copyright has been kept. 2010-02-22

For docs please refer to http://docs.icinga.org

Purpose
-------
The purpose of this addon is to allow you to send service check
results to a central monitoring server running Icinga in a secure
manner.


Contents
--------

There are two pieces to this addon:

  1) nsca

	This program runs as a daemon on the central server that runs Icinga.
	It listens for host and service check results from remote machines
	(sent using the send_nsca program described below).  Upon receiving
	data from a remote client, the daemon will make a *very* basic attempt
	at validating the data it has received from the client.  This is done
	by decrypting the data with the password stored in the nsca.cfg file.
	If the decrypted data looks okay (i.e. it was originally encrypted by
	the send_ncsa program using the same password), the daemon will make
	entries in the Icinga external command file telling Icinga to process
	the host or service check result.

	Notes: The nsca daemon must have sufficient rights to open the Icinga
	command file for writing.  Furthermore, Icinga will only process passive
	service check results that it finds in the external command file if the
	service has been defined in the host config file (i.e. hosts.cfg) and
	it is being monitored.
	Newer versions of this software will allow you to enable writing to the
	checkresult spool directory instead of the command pipe.


  2) send_nsca
	This is the client program that is used to send service check information
	from a remote machine to the nsca daemon on the central machine that runs
	Icinga. Service check information is read from the standard input in
	tab-delimited format as follows:

		<host_name>[tab]<svc_description>[tab]<return_code>[tab]<plugin_output>[newline]

	where:
		<host_name>=short name of host that the service is associated with
		<svc_description>=description of the service
		<return_code>=numeric return code
		<plugin_output>=output from service check

	Host check information is submitted in a similiar fashion - just leave out
	the service description:

		<host_name>[tab]<return_code>[tab]<plugin_output>[newline]


Compiling
---------

Most users should be able to compile the daemon and client piece with the following
commands...

./configure
make all

The binaries will be located in the src/ directory after you
run 'make all' and will have to be installed manually.



Installing
----------

The send_nsca program and associate config file (nsca.cfg) should
be placed on remote machines that you want to have communicate
with the nsca daemon.  This means that you may have to compile the
send_nsca program on the remote machine, if its not the same
OS/architecture as that of the central server.

The nsca daemon and the configuration file (nsca.cfg) should
be placed somewhere on the central server running Icinga.

NOTES:  Make sure that you specify and use the same password in
        both the nsca.cfg and send_nsca.cfg files!  If you use a
        different password to encrypt the data than you do to
	decrypt it, the nsca daemon will reject the data you send
        it.



Security
--------

There are some security implications with allowing remote clients
to provide service check results to Icinga.  Because of this, you
have the option of encrypting the packets that the NSCA client sends
to the NSCA daemon.  Read the SECURITY file for more information on
the security risks of running NSCA, along with an explanation of what
kind of protection the encryption provides you.



Running Under INETD or XINETD
-----------------------------

If you plan on running nsca under inetd or xinetd and making use
of TCP wrappers, you need to do the following things:



1) Add a line to your /etc/services file as follows (modify the port
   number as you see fit)

	nsca            5667/tcp	# NSCA



2) Add entries for the NSCA daemon to either your inetd or xinetd
   configuration files.  Which one your use will depend on which
   superserver is installed on your system.  Both methods are described
   below.  NOTE: If you run nsca under inetd or xinetd, the server_port
   and allowed_hosts variables in the nrpe configuration file are
   ignored.


   ***** INETD *****
   If your system uses the inetd superserver WITH tcpwrappers, add an
   entry to /etc/inetd.conf as follows:

	nsca    stream  tcp     nowait  <user> /usr/sbin/tcpd <nscabin> -c <nscacfg> --inetd

   If your system uses the inetd superserver WITHOUT tcpwrappers, add an
   entry to /etc/inetd.conf as follows:

	nsca 	stream 	tcp 	nowait 	<user> <nscabin> -c <nscacfg> --inetd


   - Replace <user> with the name of the user that nsca server should run as.
	Example: icinga
   - Replace <nscabin> with the path to the nsca binary on your system.
	Example: /usr/local/icinga/bin/nsca
   - Replace <nscacfg> with the path to the nsca config file on your system.
	Example: /usr/local/icinga/etc/nsca.cfg


   ***** XINETD *****
   If your system uses xinetd instead of inetd, you'll probably
   want to create a file called 'nsca' in your /etc/xinetd.d
   directory that contains the following entries (a sample config
   file called nsca.xinetd should be created in the root folder of
   the distribution after you run the configure script):


	# default: on
	# description: NSCA
	service nsca
	{
        	flags           = REUSE
	        socket_type     = stream
        	wait            = no
	        user            = <user>
		group		= <group>
        	server          = <nscabin>
	        server_args     = -c <nscacfg> --inetd
        	log_on_failure  += USERID
	        disable         = no
		only_from       = <ipaddress1> <ipaddress2> ...
	}


   - Replace <user> with the name of the user that the nsca server should run as.
   - Replace <group> with the name of the group that the nsca server should run as.
   - Replace <nscabin> with the path to the nsca binary on your system.
   - Replace <nscacfg> with the path to the nsca config file on your system.
   - Replace the <ipaddress> fields with the IP addresses of hosts which
     are allowed to connect to the NSCA daemon.  This only works if xinetd was
     compiled with support for tcpwrappers.



3) Restart inetd or xinetd will the following command (pick the
   one that is appropriate for your system:

	/etc/rc.d/init.d/inet restart

	/etc/rc.d/init.d/xinetd restart



4) Add entries to your /etc/hosts.allow and /etc/hosts.deny
   file to enable TCP wrapper protection for the nsca service.
   This is optional, although highly recommended.




Questions?
----------

If you have questions about this addon, or problems getting things
working, send an email to one of the Icinga mailing lists, visit the
portal or get on IRC.

http://www.icinga.org/support



Changelog
*********************
Icinga NSCA Changelog
*********************

3.0 - XX/XX/2012

ENHANCEMENTS
* add IPv6 support
* add config directive to submit directly to checkresults directory (Mike Lindsey) #2100
	** --with-checkresult-dir=<path> for configure
* Support multi-line check output with 4000 character limit (Mike Lindsey) #2100
* allow older clients using smaller packet size to talk to newer servers #2288
* add make install and variants, extend configure, rework init script #2111
* add --with-ext-cmd-file-dir to configure to allow specifying the path to icinga.cmd dir #2418

FIXES
* allow packets arriving with a future time stamp (Daniel Wittenberg) #2101
* fix race condition when opening command file (Matej Vela) #2400

CHANGES

* add new configure options
	** --with-checkresult-dir=<path>
	** --with-init-dir
	** --with-pid-file
	** --with-ext-cmd-file-dir=<path>
* add new make targets
	** install-daemon
	** install-client
	** install-init
	** fullinstall
	** install-daemon-config
	** install-client-config
* add new nsca.cfg options
	** check_result_path


***********************
Archived NRPE Changelog
***********************


2.8 - ??/??/????
----------------
- Added --with-log-facility option to control syslog logging (Ryan Ordway and Brian Seklecki)
- Fixed bug where daemon would segfault if mcrypt library was not initialized before timeout (Holger Weiss)
- Fixed bug with switching from dump file to command file when running under single mode (Ton Voon)
- Fix for small memory leak with running as a single process daemon (Daniel)


2.7.2 - 07/03/2007
------------------
- Fixed bug with NSCA daemon eating CPU if child process couldn't accept a connection in multi-process mode (Chris Wilson)


2.7.1 - 01/29/2007
------------------
- Fixed bug that prevented single mode daemon from working properly
- Added sample scripts for testing functionality to nsca_tests/ (Ton Voon/Altinity)


2.7 - 12/13/2006
----------------
- Fixed crash from malformed command line
- Updated to config.sub and config.guess to latest from GNU Savannah


2.6 - 04/06/2006
----------------
- Spec file fix
- Segfault fix in encryption library cleanup
- Daemon now exits with an error if it can't drop privileges
- Added chroot support (Sean Finney)
- Added support for writing a PID file
- Added support for reloading config files with SIGHUP


2.5 - 01/21/2006
----------------
- Native TCP wrapper support in daemon mode
- Memory leak fix and cleanup (Mark Ferlatte)
- Compiler warning fix (David Binderman)
- max_packet_age=0 will disable packet timestamp checks
- New config.sub and config.guess for checking local system type


2.4 - 07/23/2003
----------------
- Better support for u_int32_t detection
- Minor bug fixes


2.3 - 01/26/2003
----------------
- Minor changes to daemon init code
- Minor Makefile fixes


2.2 - 01/08/2002
----------------
- Hopefully better support for Solaris (u_int32_t)
- Syntax changes for command line arguments
- Added support for passive host checks (supported
  only in Nagios 2.x and later)
- Added sample xinetd config file (nsca.xinetd)
- Minor mods and bug fixes


2.1 - 06/10/2002
----------------
- Fix for compiling under Solaris (Scott Cokely)
- Added setuid/setgid option to config file


2.0 - 02/21/2002
----------------
- Name and version change


1.2.0 - 02/12/2002
------------------
- Compatability between older versions has been broken!
- Server now sends client IV and timestamp, fixing
  encryption problems and making it more resistent
  to "replay" attacks (Ralf Ertzinger)
- Random IV is now seeded from /dev/urandom instead
  of time() if possible
- Added directions on running under xinetd to README
- Implemented Beej's sendall() to handle incomplete send()s
- Added single-mode daemon code (David Luyer)
- Fixed problem with aggregated writes (David Luyer)
- Better autoconf detection of libmcrypt (Jay McCarthy)
- Added ability to dump check results to alternate
  dump file if command file does not exist (i.e. NetSaint
  is not running)
- Removed some old crud


1.1.2 - 09/22/2001
------------------
- Added append_to_file option to allow for opening the
  command file in either write or append mode
- Explicit integer size in packet structure
- NULL IV is used for both encryption/decryption.  This
  leaves the daemon open to "replay" attacks, but will
  disappear in future revisions when IV negotiation
  between the client and daemon is implemented.


1.1.1 - 04/26/2001
------------------
- Configure script bug fix


1.1.0 - 02/24/2001
------------------
- Added encryption routines (using libmcrypt)
- Minor code enhancements/cleanups


1.0b3 - 12/21/2000
------------------
- Removed lock file code, as 0.0.7 uses a named pipe
- Documented aggregated_writes option


1.0b2 - 08/26/2000
------------------
- Added option to use write lock when updating external
  command file


1.0b1 - 04/17/2000
------------------
- Initial cut


t/NSCATest.pm
#
# DESCRIPTION:
#	Helper object class for NSCA testing
#
# COPYRIGHT:
#	Copyright (C) 2007 Altinity Limited
#	Copyright is freely given to Ethan Galstad if included in the NSCA distribution
#
# LICENCE:
#	GNU GPLv2

package NSCATest;

use strict;
use Class::Struct;
use IO::File;

struct NSCATest => {
	config => '$',
	pid => '$',
	timeout => '$',
	};

$| = 1;	# Autoflush on

sub start {
	my ($self, $mode) = @_;
	$mode ||= "--single";

	printf "Starting nsca with $mode\n";
	system("../src/nsca -c nsca_".$self->config.".cfg $mode");

	sleep 1;	# Let daemon start
	open F, "var/nsca.pid" or die "No pid file found";
	chop(my $pid = <F>);
	close F;
	$self->pid($pid);

	open(F, "> var/icinga.cmd") or die "Cannot create var/icinga.cmd";
	close F;
	return $pid;
}

sub stop {
	my $self = shift;
	print "Stopping nsca: ".$self->pid.$/;
	kill "TERM", $self->pid;
	$self->pid(undef);
	unlink "var/icinga.cmd", "var/nsca.dump";
	sleep 1;	# Let daemon die
}

sub send {
	my ($self, $data) = @_;
	my @output = map { join("\t", @$_)."\n" } @$data;
	open SEND, "| ".$self->send_cmd;
	print SEND @output;
	close SEND;
}

sub send_cmd {
	my ($self) = @_;
	my $timeout = $self->timeout || 2;
	return "../src/send_nsca -to $timeout -p 56677 -H localhost -c send_".$self->config.".cfg";
}

sub read_cmd {
	my ($self, $file) = @_;
	$file ||= "var/icinga.cmd";
	my $fh = IO::File->new($file) or die "Can't open $file";
	$self->process_data($fh);
}

sub process_data {
	my ($self, $fh) = @_;
	my $data = [];
	while(<$fh>) {
		chop;
		my @bits = /\[\d+\] PROCESS_(?:HOST|SERVICE)_CHECK_RESULT;([^;]+);(?:([^;]+);)?([0123]);(.*)$/o;

		# Remove the service name if doesn't exist
		splice @bits, 1, 1 unless defined $bits[1];

		push @$data, [ @bits ];
	}
	return $data;
}

# Was thinking of calling $self->stop in DESTROY, but with the forking
# going on, this wouldn't work

1;


src/utils.c
/****************************************************************************
 *
 * UTILS.C - Utility functions for NSCA
 *
 * Copyright (c) 2000-2008 Ethan Galstad (nagios@nagios.org)
 * Copyright (c) 2010-2012 Icinga Development Team and Community Contributors
 *      (http://www.icinga.org)
 *
 * License: GPL
 *
 * Description:
 *
 * This file contains common unctions used in nsca and send_nsca
 *
 * License Information:
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 ****************************************************************************/

#include "../include/common.h"
#include "../include/utils.h"


/*#define DEBUG*/

static unsigned long crc32_table[256];
#ifdef HAVE_LIBMCRYPT
static volatile sig_atomic_t mcrypt_initialized = FALSE;
#endif

/* escapes newlines in a string */
char *escape_newlines(char *rawbuf) {
	char *newbuf = NULL;
	register int x, y;

	if (rawbuf == NULL)
		return NULL;

	/* allocate enough memory to escape all chars if necessary */
	if ((newbuf = malloc((strlen(rawbuf) * 2) + 1)) == NULL)
		return NULL;

	for (x = 0, y = 0; rawbuf[x] != (char)'\x0'; x++) {

		/* escape backslashes */
		if (rawbuf[x] == '\\') {
			newbuf[y++] = '\\';
			newbuf[y++] = '\\';
		}

		/* escape newlines */
		else if (rawbuf[x] == '\n') {
			newbuf[y++] = '\\';
			newbuf[y++] = 'n';
		}

		else
			newbuf[y++] = rawbuf[x];
	}
	newbuf[y] = '\x0';

	return newbuf;
}


/* build the crc table - must be called before calculating the crc value */
void generate_crc32_table(void) {
	unsigned long crc, poly;
	int i, j;

	poly = 0xEDB88320L;
	for (i = 0; i < 256; i++) {
		crc = i;
		for (j = 8; j > 0; j--) {
			if (crc & 1)
				crc = (crc >> 1) ^ poly;
			else
				crc >>= 1;
		}
		crc32_table[i] = crc;
	}

	return;
}


/* calculates the CRC 32 value for a buffer */
unsigned long calculate_crc32(char *buffer, int buffer_size) {
	register unsigned long crc;
	int this_char;
	int current_index;

	crc = 0xFFFFFFFF;

	for (current_index = 0; current_index < buffer_size; current_index++) {
		this_char = (int)buffer[current_index];
		crc = ((crc >> 8) & 0x00FFFFFF) ^ crc32_table[(crc ^ this_char) & 0xFF];
	}

	return (crc ^ 0xFFFFFFFF);
}



/* initializes encryption routines */
int encrypt_init(char *password, int encryption_method, char *received_iv, struct crypt_instance **CIptr) {
#ifdef HAVE_LIBMCRYPT
	int i;
	int iv_size;
#endif
	struct crypt_instance *CI;

	CI = malloc(sizeof(struct crypt_instance));
	*CIptr = CI;

	if (CI == NULL) {
		syslog(LOG_ERR, "Could not allocate memory for crypt instance");
		return ERROR;
	}

	/* server generates IV used for encryption */
	if (received_iv == NULL)
		generate_transmitted_iv(CI->transmitted_iv);

	/* client recieves IV from server */
	else
		memcpy(CI->transmitted_iv, received_iv, TRANSMITTED_IV_SIZE);

#ifdef HAVE_LIBMCRYPT
	CI->blocksize = 1;                      /* block size = 1 byte w/ CFB mode */
	CI->keysize = 7;                        /* default to 56 bit key length */
	CI->mcrypt_mode = "cfb";                /* CFB = 8-bit cipher-feedback mode */
	CI->mcrypt_algorithm = "unknown";
#endif

	/* XOR or no encryption */
	if (encryption_method == ENCRYPT_NONE || encryption_method == ENCRYPT_XOR)
		return OK;

#ifdef HAVE_LIBMCRYPT

	/* get the name of the mcrypt encryption algorithm to use */
	switch (encryption_method) {
	case ENCRYPT_DES:
		CI->mcrypt_algorithm = MCRYPT_DES;
		break;
	case ENCRYPT_3DES:
		CI->mcrypt_algorithm = MCRYPT_3DES;
		break;
	case ENCRYPT_CAST128:
		CI->mcrypt_algorithm = MCRYPT_CAST_128;
		break;
	case ENCRYPT_CAST256:
		CI->mcrypt_algorithm = MCRYPT_CAST_256;
		break;
	case ENCRYPT_XTEA:
		CI->mcrypt_algorithm = MCRYPT_XTEA;
		break;
	case ENCRYPT_3WAY:
		CI->mcrypt_algorithm = MCRYPT_3WAY;
		break;
	case ENCRYPT_BLOWFISH:
		CI->mcrypt_algorithm = MCRYPT_BLOWFISH;
		break;
	case ENCRYPT_TWOFISH:
		CI->mcrypt_algorithm = MCRYPT_TWOFISH;
		break;
	case ENCRYPT_LOKI97:
		CI->mcrypt_algorithm = MCRYPT_LOKI97;
		break;
	case ENCRYPT_RC2:
		CI->mcrypt_algorithm = MCRYPT_RC2;
		break;
	case ENCRYPT_ARCFOUR:
		CI->mcrypt_algorithm = MCRYPT_ARCFOUR;
		break;
	case ENCRYPT_RIJNDAEL128:
		CI->mcrypt_algorithm = MCRYPT_RIJNDAEL_128;
		break;
	case ENCRYPT_RIJNDAEL192:
		CI->mcrypt_algorithm = MCRYPT_RIJNDAEL_192;
		break;
	case ENCRYPT_RIJNDAEL256:
		CI->mcrypt_algorithm = MCRYPT_RIJNDAEL_256;
		break;
	case ENCRYPT_WAKE:
		CI->mcrypt_algorithm = MCRYPT_WAKE;
		break;
	case ENCRYPT_SERPENT:
		CI->mcrypt_algorithm = MCRYPT_SERPENT;
		break;
	case ENCRYPT_ENIGMA:
		CI->mcrypt_algorithm = MCRYPT_ENIGMA;
		break;
	case ENCRYPT_GOST:
		CI->mcrypt_algorithm = MCRYPT_GOST;
		break;
	case ENCRYPT_SAFER64:
		CI->mcrypt_algorithm = MCRYPT_SAFER_SK64;
		break;
	case ENCRYPT_SAFER128:
		CI->mcrypt_algorithm = MCRYPT_SAFER_SK128;
		break;
	case ENCRYPT_SAFERPLUS:
		CI->mcrypt_algorithm = MCRYPT_SAFERPLUS;
		break;

	default:
		CI->mcrypt_algorithm = "unknown";
		break;
	}

#ifdef DEBUG
	syslog(LOG_INFO, "Attempting to initialize '%s' crypto algorithm...", CI->mcrypt_algorithm);
#endif

	/* open encryption module */
	if ((CI->td = mcrypt_module_open(CI->mcrypt_algorithm, NULL, CI->mcrypt_mode, NULL)) == MCRYPT_FAILED) {
		syslog(LOG_ERR, "Could not open mcrypt algorithm '%s' with mode '%s'", CI->mcrypt_algorithm, CI->mcrypt_mode);
		return ERROR;
	}

#ifdef DEBUG
	syslog(LOG_INFO, "Using '%s' as crypto algorithm...", CI->mcrypt_algorithm);
#endif

	/* determine size of IV buffer for this algorithm */
	iv_size = mcrypt_enc_get_iv_size(CI->td);
	if (iv_size > TRANSMITTED_IV_SIZE) {
		syslog(LOG_ERR, "IV size for crypto algorithm exceeds limits");
		return ERROR;
	}

	/* allocate memory for IV buffer */
	if ((CI->IV = (char *)malloc(iv_size)) == NULL) {
		syslog(LOG_ERR, "Could not allocate memory for IV buffer");
		return ERROR;
	}

	/* fill IV buffer with first bytes of IV that is going to be used to crypt (determined by server) */
	for (i = 0; i < iv_size; i++)
		CI->IV[i] = CI->transmitted_iv[i];

	/* get maximum key size for this algorithm */
	CI->keysize = mcrypt_enc_get_key_size(CI->td);

	/* generate an encryption/decription key using the password */
	if ((CI->key = (char *)malloc(CI->keysize)) == NULL) {
		syslog(LOG_ERR, "Could not allocate memory for encryption/decryption key");
		return ERROR;
	}
	bzero(CI->key, CI->keysize);

	if (CI->keysize < strlen(password))
		strncpy(CI->key, password, CI->keysize);
	else
		strncpy(CI->key, password, strlen(password));

	/* initialize encryption buffers */
	mcrypt_generic_init(CI->td, CI->key, CI->keysize, CI->IV);
	mcrypt_initialized = TRUE;
#endif

	return OK;
}



/* encryption routine cleanup */
void encrypt_cleanup(int encryption_method, struct crypt_instance *CI) {

	/* no crypt instance */
	if (CI == NULL)
		return;

#ifdef HAVE_LIBMCRYPT
	/* mcrypt cleanup */
	if (encryption_method != ENCRYPT_NONE && encryption_method != ENCRYPT_XOR) {
		if (mcrypt_initialized == TRUE)
			mcrypt_generic_end(CI->td);
		free(CI->key);
		CI->key = NULL;
		free(CI->IV);
		CI->IV = NULL;
	}
#endif

	free(CI);

	return;
}



/* generates IV to use for encrypted communications (function is called by server only, client uses IV it receives from server) */
static void generate_transmitted_iv(char *transmitted_iv) {
	FILE *fp;
	int x;
	int seed = 0;

	/*********************************************************/
	/* fill IV buffer with data that's as random as possible */
	/*********************************************************/

	/* try to get seed value from /dev/urandom, as its a better source of entropy */
	fp = fopen("/dev/urandom", "r");
	if (fp != NULL) {
		seed = fgetc(fp);
		fclose(fp);
	}

	/* else fallback to using the current time as the seed */
	else
		seed = (int)time(NULL);

	/* generate pseudo-random IV */
	srand(seed);
	for (x = 0; x < TRANSMITTED_IV_SIZE; x++)
		transmitted_iv[x] = (int)((256.0 * rand()) / (RAND_MAX + 1.0));

	return;
}



/* encrypt a buffer */
void encrypt_buffer(char *buffer, int buffer_size, char *password, int encryption_method, struct crypt_instance *CI) {
	int x;
	int y;
	int password_length;

#ifdef DEBUG
	syslog(LOG_INFO, "Encrypting with algorithm #%d", encryption_method);
#endif

	/* no crypt instance */
	if (CI == NULL)
		return;

	/* no encryption */
	if (encryption_method == ENCRYPT_NONE)
		return;

	/* simple XOR "encryption" - not meant for any real security, just obfuscates data, but its fast... */
	else if (encryption_method == ENCRYPT_XOR) {

		/* rotate over IV we received from the server... */
		for (y = 0, x = 0; y < buffer_size; y++, x++) {

			/* keep rotating over IV */
			if (x >= TRANSMITTED_IV_SIZE)
				x = 0;

			buffer[y] ^= CI->transmitted_iv[x];
		}

		/* rotate over password... */
		password_length = strlen(password);
		for (y = 0, x = 0; y < buffer_size; y++, x++) {

			/* keep rotating over password */
			if (x >= password_length)
				x = 0;

			buffer[y] ^= password[x];
		}

		return;
	}

#ifdef HAVE_LIBMCRYPT
	/* use mcrypt routines */
	else {

		/* encrypt each byte of buffer, one byte at a time (CFB mode) */
		for (x = 0; x < buffer_size; x++)
			mcrypt_generic(CI->td, &buffer[x], 1);
	}
#endif

	return;
}


/* decrypt a buffer */
void decrypt_buffer(char *buffer, int buffer_size, char *password, int encryption_method, struct crypt_instance *CI) {
	int x = 0;

#ifdef DEBUG
	syslog(LOG_INFO, "Decrypting with algorithm #%d", encryption_method);
#endif

	/* no crypt instance */
	if (CI == NULL)
		return;

	/* no encryption */
	if (encryption_method == ENCRYPT_NONE)
		return;

	/* XOR "decryption" is the same as encryption */
	else if (encryption_method == ENCRYPT_XOR)
		encrypt_buffer(buffer, buffer_size, password, encryption_method, CI);

#ifdef HAVE_LIBMCRYPT
	/* use mcrypt routines */
	else {

		/* encrypt each byte of buffer, one byte at a time (CFB mode) */
		for (x = 0; x < buffer_size; x++)
			mdecrypt_generic(CI->td, &buffer[x], 1);
	}
#endif

	return;
}



/* fill a buffer with semi-random data */
void randomize_buffer(char *buffer, int buffer_size) {
	FILE *fp;
	int x;
	int seed;

	/**** FILL BUFFER WITH RANDOM ALPHA-NUMERIC CHARACTERS ****/

	/***************************************************************
	   Only use alpha-numeric characters becase plugins usually
	   only generate numbers and letters in their output.  We
	   want the buffer to contain the same set of characters as
	   plugins, so its harder to distinguish where the real output
	   ends and the rest of the buffer (padded randomly) starts.
	***************************************************************/

	/* try to get seed value from /dev/urandom, as its a better source of entropy */
	fp = fopen("/dev/urandom", "r");
	if (fp != NULL) {
		seed = fgetc(fp);
		fclose(fp);
	}

	/* else fallback to using the current time as the seed */
	else
		seed = (int)time(NULL);

	srand(seed);
	for (x = 0; x < buffer_size; x++)
		buffer[x] = (int)'0' + (int)(72.0 * rand() / (RAND_MAX + 1.0));

	return;
}



/* strips trailing newlines, carriage returns, spaces, and tabs from a string */
void strip(char *buffer) {
	int x;
	int index;

	for (x = strlen(buffer); x >= 1; x--) {
		index = x - 1;
		if (buffer[index] == ' ' || buffer[index] == '\r' || buffer[index] == '\n' || buffer[index] == '\t')
			buffer[index] = '\x0';
		else
			break;
	}

	return;
}



/* wipes an area of memory clean */
void clear_buffer(char *buffer, int buffer_length) {

	/* NULL all bytes of buffer */
	memset(buffer, '\x0', buffer_length);

	return;
}



/* show license */
void display_license(void) {

	printf("This program is free software; you can redistribute it and/or modify\n");
	printf("it under the terms of the GNU General Public License as published by\n");
	printf("the Free Software Foundation; either version 2 of the License, or\n");
	printf("(at your option) any later version.\n\n");
	printf("This program is distributed in the hope that it will be useful,\n");
	printf("but WITHOUT ANY WARRANTY; without even the implied warranty of\n");
	printf("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n");
	printf("GNU General Public License for more details.\n\n");
	printf("You should have received a copy of the GNU General Public License\n");
	printf("along with this program; if not, write to the Free Software\n");
	printf("Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n\n");

	return;
}


include/netutils.h
/************************************************************************************************
 *
 * NETUTILS.H - NSCA Network Utilities Include File
 *
 * Copyright (c) 1999-2003 Ethan Galstad (nagios@nagios.org)
 * Copyright (c) 2010-2012 Icinga Development Team and Community Contributors
 *      (http://www.icinga.org)
 *
 * License: GPL
 *
 * Description:
 *
 * This file contains common include files and function definitions used in many of the plugins.
 *
 * License Information:
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 ************************************************************************************************/

#include "../include/config.h"

int my_tcp_connect(char *,char *,int *);

int sendall(int,char *,int *);
int recvall(int,char *,int *,int);


include/common.h
/************************************************************************
 *
 * COMMON.H - NSCA Common Include File
 * Copyright (c) 1999-2003 Ethan Galstad (nagios@nagios.org)
 * Copyright (c) 2010-2012 Icinga Development Team and Community Contributors
 *      (http://www.icinga.org)
 *
 * License:
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 ************************************************************************/

#include "config.h"


#define PROGRAM_VERSION "3.0-dev"
#define MODIFICATION_DATE "10-05-2011"


#define OK		0
#define ERROR		-1

#define TRUE		1
#define FALSE		0

#define STATE_UNKNOWN  	3	/* service state return codes */
#define	STATE_CRITICAL 	2
#define STATE_WARNING 	1
#define STATE_OK       	0


#define DEFAULT_SOCKET_TIMEOUT	10	/* timeout after 10 seconds */

#define MAX_INPUT_BUFFER	5120	/* max size of most buffers we use */

/*********************** LENGTHS **********************/

/*
 * WARNING!
 * changing the lengths could cause packet failures
 * between clients and servers among different versions!
 */

#define MAX_HOST_ADDRESS_LENGTH	256	/* max size of a host address */

#define MAX_HOSTNAME_LENGTH	64
#define MAX_DESCRIPTION_LENGTH	128
#define MAX_PLUGINOUTPUT_LENGTH	4096

#define OLD_PLUGINOUTPUT_LENGTH	512
#define OLD_PACKET_LENGTH 	(( sizeof( data_packet) - ( MAX_PLUGINOUTPUT_LENGTH - OLD_PLUGINOUTPUT_LENGTH)))

#define MAX_PASSWORD_LENGTH     512

#define BLOCK_DELIMITER  	"\x17"

/********************* ENCRYPTION TYPES ****************/

#define ENCRYPT_NONE            0       /* no encryption */
#define ENCRYPT_XOR             1       /* not really encrypted, just obfuscated */

#ifdef HAVE_LIBMCRYPT
#define ENCRYPT_DES             2       /* DES */
#define ENCRYPT_3DES            3       /* 3DES or Triple DES */
#define ENCRYPT_CAST128         4       /* CAST-128 */
#define ENCRYPT_CAST256         5       /* CAST-256 */
#define ENCRYPT_XTEA            6       /* xTEA */
#define ENCRYPT_3WAY            7       /* 3-WAY */
#define ENCRYPT_BLOWFISH        8       /* SKIPJACK */
#define ENCRYPT_TWOFISH         9       /* TWOFISH */
#define ENCRYPT_LOKI97          10      /* LOKI97 */
#define ENCRYPT_RC2             11      /* RC2 */
#define ENCRYPT_ARCFOUR         12      /* RC4 */
#define ENCRYPT_RC6             13      /* RC6 */            /* UNUSED */
#define ENCRYPT_RIJNDAEL128     14      /* RIJNDAEL-128 */
#define ENCRYPT_RIJNDAEL192     15      /* RIJNDAEL-192 */
#define ENCRYPT_RIJNDAEL256     16      /* RIJNDAEL-256 */
#define ENCRYPT_MARS            17      /* MARS */           /* UNUSED */
#define ENCRYPT_PANAMA          18      /* PANAMA */         /* UNUSED */
#define ENCRYPT_WAKE            19      /* WAKE */
#define ENCRYPT_SERPENT         20      /* SERPENT */
#define ENCRYPT_IDEA            21      /* IDEA */           /* UNUSED */
#define ENCRYPT_ENIGMA          22      /* ENIGMA (Unix crypt) */
#define ENCRYPT_GOST            23      /* GOST */
#define ENCRYPT_SAFER64         24      /* SAFER-sk64 */
#define ENCRYPT_SAFER128        25      /* SAFER-sk128 */
#define ENCRYPT_SAFERPLUS       26      /* SAFER+ */
#endif



/******************** MISC DEFINITIONS *****************/

#define TRANSMITTED_IV_SIZE     128     /* size of IV to transmit - must be as big as largest IV needed for any crypto algorithm */


/*************** PACKET STRUCTURE DEFINITIONS **********/

#define NSCA_PACKET_VERSION_3   3		/* packet version identifier */
#define NSCA_PACKET_VERSION_2	2		/* older packet version identifiers */
#define NSCA_PACKET_VERSION_1	1

/* data packet containing service check results */
typedef struct data_packet_struct{
	int16_t   packet_version;
	u_int32_t crc32_value;
	u_int32_t timestamp;
	int16_t   return_code;
	char      host_name[MAX_HOSTNAME_LENGTH];
	char      svc_description[MAX_DESCRIPTION_LENGTH];
	char      plugin_output[MAX_PLUGINOUTPUT_LENGTH];
        }data_packet;

/* initialization packet containing IV and timestamp */
typedef struct init_packet_struct{
	char      iv[TRANSMITTED_IV_SIZE];
	u_int32_t timestamp;
        }init_packet;




